examples/oauth.rs

```diff
@@ -0,0 +1,245 @@
+        None => "You're not logged in.\nVisit `/auth/discord` to do so.".to_string(),
+    }
+}
+
+async fn discord_auth(Extension(client): Extension<BasicClient>) -> impl IntoResponse {
+    let (auth_url, _csrf_token) = client
+        .authorize_url(CsrfToken::new_random)
+        .add_scope(Scope::new("identify".to_string()))
+        .url();
+
+    // Redirect to Discord's oauth service
+    Redirect(auth_url.into())
+}
+
+// Valid user session required. If there is none, redirect to the auth page
+async fn protected(user: User) -> impl IntoResponse {
+    format!(
+        "Welcome to the protected area :)\nHere's your info:\n{:?}",
+        user
+    )
+}
+
+async fn logout(
+    Extension(store): Extension<MemoryStore>,
+    TypedHeader(cookies): TypedHeader<headers::Cookie>,
+) -> impl IntoResponse {
+    let cookie = cookies.get(COOKIE_NAME).unwrap();
+    let session = match store.load_session(cookie.to_string()).await.unwrap() {
+        Some(s) => s,
+        // No session active, just redirect
+        None => return Redirect("/".to_string()),
+    };
+
+    store.destroy_session(session).await.unwrap();
+
+    Redirect("/".to_string())
+}
+
+#[derive(Debug, Deserialize)]
+struct AuthRequest {
+    code: String,
+    state: String,
+}
+
+async fn login_authorized(
+    Query(query): Query<AuthRequest>,
+    Extension(store): Extension<MemoryStore>,
+    Extension(oauth_client): Extension<BasicClient>,
+) -> impl IntoResponse {
+    // Get an auth token
+    let token = oauth_client
+        .exchange_code(AuthorizationCode::new(query.code.clone()))
+        .request_async(async_http_client)
+        .await
+        .unwrap();
+
+    // Fetch user data from discord
+    let client = reqwest::Client::new();
+    let user_data: User = client
+        // https://discord.com/developers/docs/resources/user#get-current-user
+        .get("https://discordapp.com/api/users/@me")
+        .bearer_auth(token.access_token().secret())
+        .send()
+        .await
+        .unwrap()
+        .json::<User>()
+        .await
+        .unwrap();
+
+    // Create a new session filled with user data
+    let mut session = Session::new();
+    session.insert("user", &user_data).unwrap();
+
+    // Store session and get corresponding cookie
+    let cookie = store.store_session(session).await.unwrap().unwrap();
+
+    // Build the cookie
+    let cookie = format!("{}={}; SameSite=Lax; Path=/", COOKIE_NAME, cookie);
+
+    // Set cookie
+    let r = http::Response::builder()
+        .header("Location", "/")
+        .header(SET_COOKIE, cookie)
+        .status(302);
+
+    r.body(Body::empty()).unwrap()
+}
+
+// Utility to save some lines of code
+struct Redirect(String);
+impl IntoResponse for Redirect {
+    fn into_response(self) -> http::Response<Body> {
+        let builder = http::Response::builder()
+            .header("Location", self.0)
+            .status(StatusCode::FOUND);
+        builder.body(Body::empty()).unwrap()
+    }
+}
```

Comment:
We've some changes to `IntoResponse`.

I think this should work:

```suggestion
impl IntoResponse for Redirect {
    type Body = Body;
    type BodyError = type BodyError = <Self::Body as axum::body::HttpBody>::Error;
    
    fn into_response(self) -> http::Response<Body> {
        let builder = http::Response::builder()
            .header("Location", self.0)
            .status(StatusCode::FOUND);
        builder.body(Body::empty()).unwrap()
    }
}
```

Same for `AuthRedirection` below.